### ISSUE TITLE
R wrapper

### ISSUE THREAD

**Issue by dselivanov** [bigartm/bigartm#349]
I have plans to add R wrapper for BigARTM. So if somebody will come across this issue, see dselivanov/text2vec#42. Any contributions are very welcome.

**Comment by sashafrey** [comment]
@dselivanov Sounds like a good idea!

My first technical concern is around protocol buffers. Low-level API of  BigARTM relies on large number of protobuf messages, defined in https://github.com/bigartm/bigartm/blob/master/src/artm/messages.proto. Is there a way to use this technology from R?

Would it be OK to wait on this until we implement https://github.com/bigartm/bigartm/issues/325? This issue will introduce a cleaner API which will make it much easier to consume BigARTM from R. Basically we are moving down some of the features of our command-line interface (``bigartm.exe``) and python interface (``ARTM``) down into core of the library.

I expect to finish https://github.com/bigartm/bigartm/issues/325 by the end of November  (worst case mid-December). In the meantime we should discuss what's missing in BigARTM for successful integration with R.

**Comment by dselivanov** [comment]
Yes,  there is [rprotobuf](https://cran.r-project.org/web/packages/RProtoBuf/index.html) package,  and I trust authors of this package a lot. I don't have any experience with protobuf,  but believe it is not very hard. Will have a more detailed look on the next few weeks. 
It is very ok for me to wait new refactored api,  because  I also need some time to polish current text2vec functionality.

**Comment by tonytonov** [comment]
@dselivanov I was thinking about an R wrapper as well. So if you currently have an opportunity to work on it, I could help with that and maybe even find a couple of collaborators. Let me know what you think of it.

**Comment by dselivanov** [comment]
@tonytonov great. I have some initial code: https://github.com/dselivanov/bigartm/tree/master/R-pkg, but we need to review current C API. 

@sashafrey several questions:
1. Is C API stable enough? 
1. What is the most up-to-date format of protobuf messages (I mean what messages are needed and what messages are outdated)?
1. What documentation should we follow? http://docs.bigartm.org/en/stable/ref/c_interface.html seems outdated.

**Comment by ofrei** [comment]
Hi 

C API and protobuf messages are quite stable, for few good reasons - (1) it is quite expensive for us in BigARTM to maintain our own Python API if we break underlying c API, (2) there is an external C# interface on top of C interface, and we don't want to break it.

Docs for messages and C API are really bad... If you have specific questions I'd be quick to answer :)  Otherwise let me spend today and tomorrow to clean them up, and in the meantime the most reliably source of info is the code itself:

* https://github.com/bigartm/bigartm/blob/master/src/artm/messages.proto 
* https://github.com/bigartm/bigartm/blob/master/src/artm/c_interface.h 

You may also review C++ interface, which is whole contained in these two files (which are quite short --- less than 500 code lines combined):

* https://github.com/bigartm/bigartm/blob/master/src/artm/cpp_interface.h
* https://github.com/bigartm/bigartm/blob/master/src/artm/cpp_interface.cc

**Comment by ofrei** [comment]
@tonytonov @dselivanov  
(see my previous comment)

Also, good news :) We've recently upgraded BigARTM to use protobuf v3.0.0, which supports json serialization. We haven't tried it yet, but I think it should be very easy to tell c_interface that messages are json-serialized (not binary-serialized). Then you could directly craft json strings in R, and thus avoid dependency on protobuf. Should we try this?

**Comment by dselivanov** [comment]
@sashafrey thx for info, I will try... Can't promise anything, but mb next weekend.
How to tell c_interface that messages are in json (plain text), not in protobuf ?

**Comment by ofrei** [comment]
No API yet, but I can make one -- as now I know you are interested) I think the API will be like this:
``
DLL_PUBLIC int ArtmSetProtobufMessageFormatToJson(); 
DLL_PUBLIC int ArtmSetProtobufMessageFormatToBinary();
``
This will save the state (boolean flag) in c_interface, so all next API calls will treat protobuf format as JSON or Binary, for input as well as output messages. The binary will be the default, so second API is just in case, if someone needs to switch back from JSON to Binary.

**Comment by ofrei** [comment]
@dselivanov @tonytonov I made better documentation for c_interface and added support for JSON serialization (see PR #663). The documentation is here:
http://docs.bigartm.org/en/master/ref/c_interface.html
It is still from being complete --- for now I've tried to focus on how low level API is functioning, rather than documenting individual calls. The documentation for [messages.proto](http://docs.bigartm.org/en/master/ref/messages.html) is still outdated.
Please ask me questions if you need more info on specific APIs or messages, and eventually I'll add this to documentation or [Q&A](https://github.com/bigartm/bigartm/wiki/Q&A)
